neo/core/dataobject.py

```diff
@@ -0,0 +1,365 @@
+            # In case the user wants this, the list needs to be converted
+            #  to np.ndarray first
+            for element in value:
+                if not isinstance(element, dtype):
+                    raise ValueError("Lists with different types are not supported for "
+                                     "array annotations. ")
+
+            # Create arrays from lists, because array annotations should be numpy arrays
+            try:
+                value = np.array(value)
+            except ValueError as e:
+                msg = str(e)
+                if "setting an array element with a sequence." in msg:
+                    raise ValueError("Scalar Quantities and array Quanitities cannot be "
+                                     "combined into a single array")
+                else:
+                    raise e
+
+    return value
+
+
+class DataObject(BaseNeo, pq.Quantity):
+    '''
+    This is the base class from which all objects containing data inherit
+    It contains common functionality for all those objects and handles array_annotations.
+
+    Common functionality that is not included in BaseNeo includes:
+    - duplicating with new data
+    - rescaling the object
+    - copying the object
+    - returning it as pq.Quantity or np.ndarray
+    - handling of array_annotations
+
+    Array_annotations are a kind of annotations that contain metadata for every data point,
+    i.e. per timestamp (in SpikeTrain, Event and Epoch) or signal channel (in AnalogSignal
+    and IrregularlySampledSignal).
+    They can contain the same data types as regular annotations, but are always represented
+    as numpy arrays of the same length as the number of data points of the annotated neo object.
+    '''
+
+    def __init__(self, name=None, description=None, file_origin=None, array_annotations=None,
+                 **annotations):
+        """
+        This method is called from each data object and initializes the newly created object by
+        adding array annotations and calling __init__ of the super class, where more annotations
+        and attributes are processed.
+        """
+
+        if not hasattr(self, 'array_annotations') or not self.array_annotations:
+            self.array_annotations = ArrayDict(self._get_arr_ann_length())
+        if array_annotations is not None:
+            self.array_annotate(**array_annotations)
+
+        BaseNeo.__init__(self, name=name, description=description,
+                         file_origin=file_origin, **annotations)
+
+    def array_annotate(self, **array_annotations):
+
+        """
+        Add annotations (non-standardized metadata) as arrays to a Neo data object.
+
+        Example:
+
+        >>> obj.array_annotate(code=['a', 'b', 'a'], category=[2, 1, 1])
+        >>> obj.array_annotations['code'][1]
+        'b'
+        """
+
+        self.array_annotations.update(array_annotations)
+
+    def array_annotations_at_index(self, index):
```

Comment:
I think in future versions it would make sense to also support slices as index here